[IRGen] Fix crashes involving ObjC generic params

beccadax opened the pull request:
Sema allows you to pass type-pinning parameters into an extension of an Objective-C generic class, but IRGen did not properly handle erasing these types to existential types in runtime metadata. This commit corrects that mistake.

Fixes rdar://80708693.

beccadax committed:
[IRGen] Accommodate ObjC generic classes better

Classes using the type-erased Objective-C generics model are represented in parts of IRGen as UnboundGenericTypes, which is a problem because a number of code paths expect all generic types to be bound. Update some of these that are involved in extensions on ObjC generic types.

slavapestov commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -85,6 +85,12 @@ llvm::cl::opt<bool> UseBasicDynamicReplacement(
 
 namespace {
+
+ClassDecl *getClassBoundOrNot(CanType ty) {
```
Suggestion: `dyn_cast_or_null<ClassDecl>(ty->getAnyNominal())`

slavapestov commented on lib/IRGen/MetadataRequest.cpp:
```diff
@@ -444,14 +444,29 @@ static llvm::Value *emitObjCMetadataRef(IRGenFunction &IGF,
     }
+    // Erase ObjC generic parameters to an equivalent existential.
+    if (auto archT = t->getAs<PrimaryArchetypeType>()) {
+      if (isObjCGenericParam(archT->getInterfaceType()->getDecl())) {
```
I'd like to get rid of the backpointer to the GenericTypeParamDecl from GenericTypeParamType one day since it causes problems all over the place. It's not the end of the world to add a new usage here, but is there any other way you can think of expressing this?

beccadax replied:
I think we'd need a bit on `GenericTypeParamType` saying that it's type-erased. I could add an `isTypeErased()` accessor to it that's currently implemented in terms of `getDecl()`, but could be changed to work from a flag bit in the future.

beccadax replied:
(The only other approach I could come up with was to make the ObjC-generic-ness a feature of the `GenericSignature` and therefore known from the `PrimaryArchetypeType`, but that seems like it'd wind through a ton of code and I don't see a good way to fit objc-generic-ness into the existing model.)

slavapestov replied:
can we look at the nominal type whose metadata we're instantiating? Ie, if you have `Foo<T>`, you're erasing `T` to some protocol or class iff `Foo` is an Objective-C generic, correct?

slavapestov replied:
... and if Foo _is not_ an Objective-C lightweight generic class, then erasing the generic parameter to its existential bound is not valid in general, because existentials do not self-conform.

beccadax replied:
We don't have `Foo<T>` here; if we did, the existing case (converting it to an unbound type for `Foo`) would cover it. We have something like `T` or `T.Type` here, typically because we wrote an extension method with a `T` or `T.Type` parameter. So what I'm trying to handle is a type like this:

```
(metatype_type thick
  (primary_archetype_type address=0x7fdf2d595f18 class layout=AnyObject name=ObjectType))
```

I haven't been able to come up with a way to go from the `PrimaryArchetypeType` to the declaration the generic parameter belongs to without going through `getInterfaceType()->getDecl()`, and I don't really have any other context to work from in this method. Maybe if I moved `getRuntimeReifiedType()` to IRGenFunction, I'd be able to see if I'm in a pseudogeneric function. But isn't this really more a property of the parameter or the generic signature used in the archetype? Surely generic parameters for other types also pass through this function.

slavapestov replied:
What happens if you have a canonical type (getDecl() is null)?

slavapestov replied:
Checking at a higher level if we're inside of a pseudogeneric function sounds like a good idea. Maybe emitArchetypeTypeMetadataRef() is the right place for this? Also would this subsume the existing erasure to UnboundGenericType? It would be nice to eliminate that as well.

beccadax commented:
@swift-ci test

beccadax committed:
[NFC] Rearrange irgen::MetadataSource initializers

Preparing to accommodate additional layouts for MetadataSource.

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -3013,7 +3013,8 @@ Type ArchetypeType::getExistentialType() const {
   return ProtocolCompositionType::get(
-     const_cast<ArchetypeType*>(this)->getASTContext(), constraintTypes, false);
+     const_cast<ArchetypeType*>(this)->getASTContext(), constraintTypes,
+                                      requiresClass());
```
Oh, nice catch!

slavapestov commented on lib/IRGen/MetadataRequest.cpp:
```diff
@@ -453,12 +453,6 @@ CanType IRGenModule::getRuntimeReifiedType(CanType type) {
     if (CanType(t).isTypeErasedGenericClassType()) {
       return t->getAnyNominal()->getDeclaredType()->getCanonicalType();
```
Do you still need this?

beccadax replied:
Yes, unfortunately.

beccadax committed:
[NFC] Unify helpers for checking for ObjC generics

There were two differently-named and slightly differently-implemented checks on nominal type decls. Clean this up a bit.

beccadax committed:
[IRGen] Fix crashes involving ObjC generic params

Sema allows you to pass type-pinning parameters into an extension of an Objective-C generic class, but IRGen did not properly handle erasing these types to existential types in runtime metadata. This commit corrects that mistake.

beccadax commented:
Force-pushed an expungement of the original implementation (although I kept the changes correcting the `isTypeErasedGenericClass()`/`usesObjCGenericModel()` split, since that was just silly).

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test linux platform

beccadax commented:
The Linux failure is unrelated and apparently caused by a recent change to SwiftPM to use Swift Collections.

beccadax commented:
@swift-ci please test linux platform
